File: 200-number-of-islands/200-number-of-islands.cpp
class Solution {
public:
    
    void dfs(vector<vector<char>>& grid, int c, int r, vector<vector<bool>>& vis){
        if(c>=grid.size() || c<0 || r>=grid[0].size() || r<0 || vis[c][r]==1 || grid[c][r]!='1')
            return;
        
        vis[c][r] = 1;
        dfs(grid, c+1, r, vis);
        dfs(grid, c, r+1, vis);
        dfs(grid, c-1, r, vis);
        dfs(grid, c, r-1, vis);
    }
    
    
    
    
    int numIslands(vector<vector<char>>& grid) {
        int ans = 0;
        int m = grid.size();
        int n = grid[0].size();
        vector<vector<bool>> vis(m,vector<bool>(n,0));
        for(int i=0 ; i<m ; i++){
            for(int j=0 ; j<n ; j++){
                vis[i][j] = 0;
            }
        }
        
        
        for(int i=0 ; i<m ; i++){
            for(int j=0 ; j<n ; j++){
                if(vis[i][j]==0 && grid[i][j]=='1'){
                    ans++;
                    dfs(grid, i, j, vis);
                }
            }
        }
        
        return ans;
    }
};

File: 200-number-of-islands/200-number-of-islands.java
class Solution {
    public int numIslands(char[][] grid) {
        
        
        int[][] vis = new int[grid.length][grid[0].length];
        int res = 0;
        for(int i=0 ; i<grid.length ; i++)
            for(int j=0 ; j<grid[0].length ; j++){
                if(vis[i][j]!=1 && grid[i][j]=='1'){
                    dfs(grid, i, j, vis);
                    res++;
                }
            }
        
        return res;
    }
    
    public void dfs(char[][] grid, int x, int y, int[][]vis){
        if(x<0 || x>=grid.length || y<0 || y>=grid[0].length || vis[x][y]==1)
            return;
        
        if(grid[x][y] == '0')
            return;
        
        vis[x][y] = 1;
        dfs(grid, x+1, y, vis);
        dfs(grid, x-1, y, vis);
        dfs(grid, x, y+1, vis);
        dfs(grid, x, y-1, vis);
   
    }
}
